Added GTIN and Condition to variant for structured data use

rahulsingh321 opened the pull request:
# Introduction of Structured Data for Item Condition and Global Trade Item Number ("GTIN"). 

**Contained Information**
- Brief Introduction into structured data
- Detailed Description of added fields including 3rd party support
- Screenshots of Admin Interface
- Example of API request to post / get / update an item with the fields

# Introduction
This is the first in a series of PRs to implement full structured data support on Solidus.
Structured Data is the common standard to exchange product and content information via the frontend with search engines and marketing platforms and is defined by the schema.org organisation (through the founding companies Google, Microsoft, Yahoo and Yandex).

It has become the defacto standard to communicate with search engines and also many social networks (such as Facebook).

In this first PR following fields are tackled: 
- GTIN
- ItemCondition

This scope of this PR is not to provide structured data on the frontend but merely to provide the information required to generate fully compliant structured data on frontend and feeds without 3rd party extensions or added fields.

## GTIN (frequently also known as UPC / EAN in Europe) 
### Definition
A unique identifier for products usually given by either the manufacturer or the official distributor of a product. Frequently required to have stock listed with marketplaces. It's usually used to identify a product uniquely across multiple steps in the supply chain. While the GTIN is a unique identifier for a product entity it is in no mean a serial number (eg every iPhone of the same model and configuration sold in one country share the UPC, but not the serial number).
It's frequently used to match products in stock synchronizations or order APIs. 

### Support 
#### Search Engines
| Search Engine | Supported                     |
|---------------|-------------------------------|
| Google        | YES                           |
| Bing          | YES                           |
| Yahoo         | YES                           |
| Yandex        | YES                           |
| Baidoo        | Partial / Closed Beta Program |

#### Marketplaces
| Search Engine | Supported                                 |
|---------------|-------------------------------------------|
| Amazon        | YES |
| Facebook      | YES  |
| Ebay          | YES |
| Etsy          | No                                        |
| Vinted        | No                     |

#### Search Ads
| Search Engine | Supported                     |
|---------------|-------------------------------|
| Google Shopping       | YES                           |
| Bing          | YES (mandatory if given by manufacturer)                          |
| Yahoo         | YES                           |
| Yandex        | YES                           |
| Baidoo        | Unknown |

## Item Condition 

### Definition
The item condition has found wide acceptance over the past years to annotate if a product is 

1. New
2. Used
3. Refurbished
4. Damaged

While the origins where originally found in second hand marketplaces and the car industry, it became widely used with the emergence of the circular economy. 

### Support
#### Search Engines
| Search Engine | Supported                     |
|---------------|-------------------------------|
| Google        | YES                           |
| Bing          | YES                           |
| Yahoo         | YES                           |
| Yandex        | YES                           |
| Baidoo        | Partial / Closed Beta Program |

#### Marketplaces
| Search Engine | Supported                                 |
|---------------|-------------------------------------------|
| Amazon        | YES (in addition Collectable)             |
| Facebook      | YES (in addition various usage grades)    |
| Ebay          | YES (with various additional information) |
| Etsy          | No                                        |
| Vinted        | YES (Free field)                          |

#### Search Ads
| Search Engine | Supported                     |
|---------------|-------------------------------|
| Google Shopping       | YES                           |
| Bing          | YES (optional)                          |
| Yahoo         | YES                           |
| Yandex        | YES                           |
| Baidoo        | No |

Concerns: #6060

# Screenshots Admin Backend 
## Product
![Screenshot from 2025-02-02 21-50-59](https://github.com/user-attachments/assets/025d1044-b58d-4cdc-86ae-6aa9123e9b12)

## Variant 
![Screenshot from 2025-02-02 21-50-41](https://github.com/user-attachments/assets/b32dac63-7cc7-4a14-a02a-c161d910a855)

# APIs
## Get Product 
![Screenshot-from-2025-02-02-22-33-10-02-02-2025_10_33_PM](https://github.com/user-attachments/assets/acdd8858-a2cb-4261-9ffa-fd3ea9428b00)
## Update Product
![Screenshot-from-2025-02-02-23-02-24-02-02-2025_11_02_PM](https://github.com/user-attachments/assets/5e162a33-30a7-496b-b1d6-c0dc23d1c09f)


# Possible Improvements 

- [ ] Validate GTIN against current schemes (EAN 9, UPC, ...)

mamhoff committed:
Remove Spree::UserAddress#archived flag

This flag is not very useful and gives rise to bugs.

It is not very useful, as we only need it to "resurrect" addresses that
the user inputs newly, but that match an address that has been removed
from the address book in the past for some reason. Why not remove the
record from the join table directly? Much simpler, and less code to worry about.

The bug we found that exists only because of the existence of this
column is the following: If an address is added to the address book that
matches an existing user address that is archived, there is a chance of
it throwing a uniqueness validation error, because the uniqueness check
on user/address on the Spree::UserAddress model does not take into
account the "archived" flag.

The third reason why I think this should go is for data protection
reasons: If I call "delete" on a record, I want it to be gone, not
half-gone. And in the case of a join table representing which addresses
a user wants to be presented with again, I very much want it gone.

Not archived.

Note also: Removing this complexity removes explanatory comments in
otherwise straight-forward API endpoints.

mamhoff committed:
Add getter for configured class names

When configuring Flickwerk, we need a way to get at class names without
constantizing (i.e. loading) those classes. This adds some specs for the
`class_name_attribute` class method on
Spree::Preferences::Configuration, and generates a new getter that has
`_name` attached to the method to get just the string name.

mamhoff committed:
Move decorators to patches/, configure Flickwerk

This stops us from loading line item, order, adjustment, shipment,
shipping rate, and the order recalculator on app startup.

mamhoff committed:
Legacy Promotions: Use Flickwerk for patches

This should stop us from loading order, line items, adjustment,
calculators, the order updater, product, and shipment on app startup.

mamhoff committed:
Move line_item_comparison_hooks config to Spree::Config

Setting line item comparison hooks in a `config.to_prepare` block will
currently autoload the Spree::Order model (and subsequently any model
referenced in that class definition), slowing down the boot process in
development mode (when we want our app/ directory to be autoloaded.

This moves setting those hooks to Spree::Config, which is already loaded
by the time initializers run. This allows us to move code out of
`config.to_prepare` hooks, and stop the order class from being
inadvertently eager loaded.

mamhoff committed:
Extract ClassMethods constant into autoloadable file

Otherwise, patching it does not work.

jarednorman committed:
Refactor line item total calculations

While working on the in-memory updater in #5872, we found the need to
change how item totals were being calculated, so that we could mark
adjustments for destruction without actually destroying them, while
still keeping tax adjustments intact. This change is completely
backwards-compatible with the current OrderUpdater, so to reduce the
scope of our PR, we wanted to make this change separately.

Since the OrderUpdater is already very large, this helps reduce its
responsibilities and makes it easier to test this behaviour. We don't
see it as necessary to make this a configurable class, but this change
leaves that option open in the future.

Co-authored-by: Adam Mueller <adam@super.gd>
Co-authored-by: Alistair Norman <alistair@super.gd>
Co-authored-by: Chris Todorov <chris@super.gd>
Co-authored-by: Harmony Bouvier <harmony@super.gd>
Co-authored-by: Sofia Besenski <sofia@super.gd>
Co-authored-by: benjamin wil <benjamin@super.gd>

adammathys committed:
Make ItemTotal class configurable

I don't know if there's all that much customizing people will perform on
this class, but it's easy enough to make it configurable just in case.

fthobe commented on backend/app/views/spree/admin/products/_form.html.erb:
```diff
@@ -173,7 +173,7 @@
       <% else %>
         <div id="shipping_specs" class="row">
-          <% [:height, :width, :depth, :weight].each_with_index do |field, index| %>
+          <% [:height, :width, :depth, :weight, :gtin].each_with_index do |field, index| %>
```
Gtin is an identifier and not a shipping spec, it belongs close to other identifiers such as SKU

fthobe replied:
FYI: 
GTIN = Global Trade Item Number

You know them, for example EAN is a GTIN, they look like this: 
![image](https://github.com/user-attachments/assets/0f67f87a-481b-4ad6-b5d2-999c76ab6b67)

fthobe commented on core/app/models/spree/variant.rb:
```diff
@@ -28,6 +28,8 @@ class Variant < Spree::Base
     attr_writer :rebuild_vat_prices
     include Spree::DefaultPrice
 
+    CONDITIONS = %w[New Refurbished Used Damaged].freeze
```
Enumeration members
- DamagedCondition
- NewCondition
- RefurbishedCondition
- UsedCondition

Different platforms requiere different visualization of condition: 
For new following values are accepted: 

1. Google: `NewCondition` or `https://schema.org/NewCondition`
2. Facebook:  `https://schema.org/NewCondition`
3. Bing: `New`
4. Yandex: `https://schema.org/NewCondition`

We will later on offer this data in the various formats through feeds and / or structured data on the single product and category pages.

fthobe replied:
Please add following comment: 
`# Consult the various references for structured data and validate frontend-output where neccessary`

fthobe reviewed (commented):
Please take a look

JustShah committed:
Add concern and validations for metadata with config options for keys and length

Introduced validations and configurable limits for metadata attributes
(`customer_metadata` and `admin_metadata`) to ensure consistency and prevent
overly large or complex structures.

**Implementation:**
- Set default values for `customer_metadata` and `admin_metadata` to empty hashes (`{}`) for proper initialization.
- Implemented validation to limit the number of keys, and the length of keys and values in metadata.
- Added configuration checks to ensure metadata does not exceed the limits defined in `Spree::Config`.
- Introduced preferences: `meta_data_max_keys`, `meta_data_max_key_length`, and `meta_data_max_value_length` to control metadata size:
  - `meta_data_max_keys`: Max keys allowed (default: 6).
  - `meta_data_max_key_length`: Max key length (default: 16).
  - `meta_data_max_value_length`: Max value length (default: 256).
- Created a shared concern to handle metadata validation logic.
- Added tests to ensure default values and validation logic work correctly.
- Created a shared example that can be used in other specs to ensure consistent testing of metadata validation across various parts of the system.

JustShah committed:
Include concern to models

Call the metadata concern to various models to validate and add admin and customer metadata

JustShah committed:
Add restriction for metadata update after order is complete

User cannot update customer_metadata after the status of order is complete

JustShah committed:
Add permalink history for taxon on friendly-id

- Introduced FriendlyId to the Taxon model to enable slug history tracking.
- Replaced before_create and before_update callbacks with a before_save callback for better slug management.
- Implemented a custom should_generate_new_friendly_id? method to control when a new slug is generated, specifically when the permalink changes.
- Added a custom normalize_friendly_id method to ensure proper slug formatting for nested taxons.
- Included a test to verify that the history of permalinks is correctly stored when a taxon's permalink is updated, ensuring both initial and updated slugs are retained.

This change enhances the flexibility and reliability of handling slugs for taxons.

fthobe commented:
@rahulsingh321 please publish as draft till it's ready for review (tests pass)

fthobe committed:
Fixed wording, typos, license notice & linked to migration guide

Following changes have been made: 
- given that the license is identical to parent repository, licensed under was changed
- integrated link to migration guide
- fixed some wording to give a more positive drive
- fixed some typos

fthobe committed:
Update README.md

forkata committed:
Fix field container identifier on admin stock location

The `address1` container incorrectly had the same identifier as the
"Check stock on transfer" input checkbox. This fixes the typo, so this
container can be referenced in Deface overrides, etc.

Co-authored-by: Benjamin Willems <benjamin@super.gd>

tvdeyen committed:
Merge pull request #3852 from mamhoff/remove-archived-user-addresses

Remove Spree::UserAddress#archived flag

elia committed:
Merge pull request #6106 from fthobe/main

Fixed wording, typos, license notice & linked to migration guide

tvdeyen committed:
Merge pull request #6100 from gms-electronics/add-permalink-history-to-taxon

Add permalink history for taxon on friendly-id

chaimann committed:
Use modal UI for legacy_promotions index/add/edit

Introduces changes to legacy promotions UI based on https://github.com/solidusio/solidus/pull/6046

chaimann committed:
Use modal UI for solidus_promotions index/add/edit

Introduces changes to new solidus promotions UI
based on https://github.com/solidusio/solidus/pull/6046.

Had to override few methods in controller and components
due to differences in routing and naming in solidus_promotions.

Copied capybara driver setup from legacy_promotions.

tvdeyen committed:
Merge pull request #6107 from SuperGoodSoft/fix-typo-in-stock-location-form-partial

Fix field container identifier on admin stock location

chaimann committed:
Add legacy_promotions and solidus_promotions calls to bin/build

chaimann committed:
Filter shared_examples from coverage

jarednorman committed:
Merge pull request #6080 from SuperGoodSoft/refactor-line-item-total-updates

Refactor Line Item Total Calculations

tvdeyen committed:
CI: Test with Ruby 3.4

tvdeyen committed:
Add csv as dependency to promotions gems

Since Ruby 3.4 the csv library is it's own gem

chaimann committed:
Fix missing options in select tags

A small improvement on changes from https://github.com/solidusio/solidus/pull/6091
where in Rails 7 calling `tag.method(@tag).parameters.any? { |_type, name| name == :content }`
would always return false when @tag == :select, since `tag.select` method's
signature has only `*` (:rest) as its parameter, thus not passing any `<option>`s for select tag.

Somehow the problem was not being flagged by CI builds, but admin specs failed locally.

Instead, use explicit void elements list to understand whether given element is void, therefore does not accept `content`.

JustShah committed:
Introduce primary taxon for product

Created a new association between Product and Primary Taxon using the existing Taxon model.

JustShah committed:
Enable API support for primary taxon

Allows access to the newly added attribute through existing product APIs.

JustShah committed:
Admin form support for primary taxon

Updated the product form for primary taxon and modified the related jQuery to display and select a specific taxon.

tvdeyen committed:
DummyGenerator: Remove unused script/rails file

This file gets copied into the wrong folder and is not
used by Rails since decades.

tvdeyen committed:
DummyGenerator: Update application template

We do not have production group in generated dummy apps
and the syntax is outdated. This is what the Rails engine
generator does as well. So let's stick with the framework

tvdeyen committed:
DummyGenerator: Add Sprockets manifest

Rails stopped creating sprockets files but we still use sprockets. And we need to manifest so that test apps
generated during specs do not fail with a sprockets error.

chaimann committed:
Fix rubocop offence

JustShah committed:
Introduce Customer and Admin Metadata Restriction Preference

This configuration adds the ability to enable or disable metadata restrictions as needed.

Following configurations are possible:
- `meta_data_validation_enabled` : Enables or disables restrictions globally.

tvdeyen committed:
Merge pull request #6121 from blish/fix-dummy-generator-sprockets

Fix DummyApp Generator

tvdeyen committed:
Fix log entry spec for Ruby 3.4

Ruby 3.4 finally adds the new has key syntax into inspect.
Since we still support older Ruby versions we just remove
the expectation of the hash key.

mamhoff committed:
Rails 8: Include manifest.js in install generator

We found we need to do this for dummy apps, but we also need to do it
for generated apps.

tvdeyen committed:
Merge pull request #6122 from mamhoff/create-manifest-js-in-generated-apps

Rails 8: Include manifest.js in install generator

tvdeyen committed:
Merge pull request #6117 from blish/ruby-3.4

Add Ruby 3.4 support

tvdeyen committed:
Merge pull request #6109 from gms-electronics/add-primary-breadcrumb-to-product

Add primary Taxon to products (#6047)

mamhoff committed:
Revert "Merge pull request #6122 from mamhoff/create-manifest-js-in-generated-apps"

This reverts commit bec42f64b9008dda63dc89c69de68fdc0076f293, reversing
changes made to 37a5c8a3182e58185a2ec5691a47721c320ad764.

kennyadsl committed:
Merge pull request #6120 from chaimann/fix-select-component-missing-content

Fix missing options in select tags

chaimann committed:
Remove Display Order from side menu in new admin

Removes "Display Order" menu item from side menu as per
https://github.com/orgs/solidusio/projects/12/views/1?pane=issue&itemId=52590353.

Note: if enabled in initializer, `config.import_menu_items_from_backend!`
will still put "Display order" menu item since it is present in legacy
backend menu.

tvdeyen committed:
Merge pull request #6101 from chaimann/admin-promotion-categories-add-edit

Admin promotion categories add/edit

tvdeyen committed:
Merge pull request #6050 from mamhoff/configure-line-item-comparison-hooks-in-config

Move line_item_comparison_hooks config to Spree::Config

chaimann committed:
Fix migration typo (missing keyword `foreign_key`)

Also set `primary_taxon_id` column type to be integer to match `spree_taxons.id`

JustShah committed:
Linting

JustShah committed:
Add metadata columns for resources

This migration adds customer_metadata and admin_metadata to various tables

JustShah committed:
Add admin_metadata to attributes for admins

Introduce the `admin_metadata` attribute to support dynamic metadata handling for admins.
It enables configuration of metadata via API preferences, and ensures that metadata fields
are properly permitted for admin users through new helper methods.

**Implementation:**
- Added `admin_metadata` to resources for admins.
- Integrated `metadata_permit_parameters` and `metadata_api_parameters preference` in API configuration.
- Implemented `permitted_<resource>_attributes` and `<method_name>_attributes` methods for dynamic metadata permissions.
- Added `extract_metadata` in `line_items_controller` to handle metadata fields.

JustShah committed:
Permit `customer_metadata` attributes for APIs and add support for metadata handling

Update the API configuration to permit the `customer_metadata` attribute across various models,
such as orders, payments, shipments, and return authorizations. This change allows better handling of customer metadata within API requests,
ensuring consistent storage and validation of metadata attributes. Additionally, it updates the `preference` configurations for models to include `customer_metadata`.

**Implementation:**
- Added `customer_metadata` to the permitted attributes for several API resources.
- Updated various `preference` settings, including `order_attributes`, `payment_attributes`, `shipment_attributes`, etc., to include `customer_metadata`.
- Ensured that `customer_metadata` is correctly permitted in the API parameters for each relevant model, allowing admins and users to store custom metadata.
- This change improves the flexibility of metadata management within the API, enabling dynamic handling and easier customizations.

tvdeyen committed:
Merge pull request #6118 from gms-electronics/metadata-draft-pr

Admin and User Metadata for transactional ressources and users (#5897)

tvdeyen committed:
Add sprockets manifest before running rails commands

Due to a bug in sprockets-rails[1] we need to manually
add the sprockets manifest into the generated rails
app before running any rails commands inside the rails
app folder.

[1](https://github.com/rails/sprockets-rails/pull/546)

tvdeyen committed:
Merge pull request #6124 from mamhoff/remove-manifest-generation

Revert "Merge pull request #6122 from mamhoff/create-manifest-js-in-g…

tvdeyen committed:
Merge pull request #6126 from chaimann/fix-migration-typo

Fix migration typo (missing keyword `foreign_key`)

mamhoff committed:
Rubocop: Ignore Rails/Delegate

This rule makes code shorter, but much less understandable.

tvdeyen committed:
Merge pull request #6119 from chaimann/remove-display-order-from-nav

Remove Display Order from side menu

tvdeyen committed:
Merge pull request #6131 from mamhoff/ignore-new-rubocop-rule

Rubocop: Ignore Rails/Delegate

tvdeyen committed:
Merge pull request #6049 from mamhoff/flickwerk-promotions

Use Flickwerk for loading patches in solidus_promotions and solidus_legacy_promotions

tvdeyen committed:
Merge pull request #6130 from blish/add-not-rails-8-install

Add not about sprockets manifest before running rails commands

tvdeyen committed:
Fix sandbox command for Rails 8

Rails 8 does not create a sprockets manifest anymore,
but we still depend on it. Since rails commands do not
work anymore, because sprockets raises an error if no
manifest.js file is present we need to create it in
order to be able to run rails new

rahulsingh321 committed:
Adding Conditions and GTIN to products and variant

Summary

This commit adds GTIN and Conditions to products to allow a
more efficient integration of structured data and the related
services such as Google Shopping or Facebook Marketplaces.

Changes

- Add Conditions and GTIN to products and variants
- Added the fields to the Admin Interface
- Add the values to API
- Provides samples for GTIN (format EAN 8) and Condition (New) to all
sample products.
- Added Testing of the related fields
- feat(i18n): Differentiate GTIN and Condition labels for product and
variant forms.

swamp09 committed:
[Admin] Fix tailwindcss-rails Version to v3 for Solidus Admin Compatibility

## Background

Currently, running `bin/rails solidus_admin:tailwindcss:install` installs the latest v4 of `tailwindcss-rails`. This causes compatibility issues because Solidus Admin expects the Tailwind CSS v3 configuration.

Tailwind CSS v4 introduces CSS First Configuration, which alters the structure and handling of the `tailwind.config.js` file. Using v4 results in a configuration where `tailwind.config.js` does not function properly. Reference: https://tailwindcss.com/blog/tailwindcss-v4#css-first-configuration

While ideally, Solidus Admin should be updated to work with Tailwind CSS v4 and later, we have restricted it to v3 for now to ensure correct functionality. Additionally, the maintainers of `tailwindcss-rails` do not recommend upgrading to v4 at this time (see: https://github.com/rails/tailwindcss-rails?tab=readme-ov-file#you-dont-_have_-to-upgrade).

## Solution

Modified the installation command in `install_tailwindcss.rb` to explicitly install v3:

```ruby
unless bundle_command "show tailwindcss-rails"
  bundle_command "add tailwindcss-rails --version '~>3.0'"
end
```

This change ensures that running bin/rails solidus_admin:tailwindcss:install installs Tailwind CSS v3, which is compatible with Solidus Admin's current configuration.

If there is anything else that should be added or any improvements that can be made, please let me know.

tvdeyen committed:
Merge pull request #6135 from swamp09/fix_tailwindcss_rails_v3_to_install_tailwindcss

[Admin] Fix tailwindcss-rails Version to v3 for Solidus Admin Compatibility

tvdeyen committed:
Merge pull request #6133 from blish/fix-sandbox-rails8

Fix sandbox command for Rails 8

tvdeyen committed:
Update solidus-api.oas.yml

Reflect the changes to the conditions values.
